storm-server/src/main/java/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -1288,6 +1296,18 @@ public DynamicState withPendingChangingBlobs(Set<Future<Void>> pendingChangingBl
                                     pendingChangingBlobs,
                                     pendingChangingBlobsAssignment, this.slotMetrics);
         }
+
+        private void cancelPendingBlobs() {
+            // Make sure any downloading blobs in the background are stopped.
+            // This prevents a race condition where we could be adding references to a
+            // delayed downloading blob after the slot gets released, causing orphaned blobs.
+            for (Future future : pendingChangingBlobs) {
+                if (!future.isDone()) {
+                    LOG.info("Canceling download of {}", future);
+                    future.cancel(true);
+                }
+            }
+        }
```

Comment:
Does this pendingChangingBlobs list need to be cleared before exiting method?

Reply:
This causes various asserts in the integration tests.  I tried creating new DynamicStates without any futures, and that caused other problems.  This is beyond my understanding, but having the futures canceled should prevent the race conditions.